### PR TITLE
Use address, undefined behaviour, and leak sanitizers in CI

### PR DESCRIPTION
**Pull request by msimberg**
This adds a pipeline with address, undefined behaviour, and leak sanitizer. Leak sanitizer is enabled by default with address sanitizer (and explicitly with `detect_leaks=1`). I've added exactly one pipeline which uses clang 15 (the newest clang available in the currently used base image) with `RelWithDebInfo` (debug symbols are useful for reporting errors, but so are optimizations to make sure we run something at least slightly resembling a release build).

As far as I can tell address sanitizer reported no failures. The option `detect_stack_use_after_return` is explicitly disabled (it's also disabled by default) because it causes errors that seem to be failures within address sanitizer itself. These failures also appear in pika's CI where I've also disabled the option. The failures may point to a real issue in pika, but I don't know yet what so I'm keeping the option off to get something useful into CI.

pika also has an option to reduce the chances of false positives with sanitizers but this is not yet in the spack package. I will enable this option in a follow-up PR.

Undefined behaviour sanitizer reported integer overflows in the `test_util_math` test, which I've changed to avoid overflows.

I've disabled `libSegFault` for the sanitizer builds since the sanitizers will report segmentation faults anyway.

### CODE REVIEW

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
This is now ready for review.

I did some intermediate broken commits to check that the sanitizers actually report failures. On https://github.com/eth-cscs/DLA-Future/commit/795dd287a05e8091456912a780af8844f08d04cf address sanitizer easily detects the out-of-bounds access: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/jobs/6217408291#L5669. It's obviously not perfect and isn't going to catch every mistake, but hopefully it'll help catch the simple ones at least.

**Comment by msimberg** [comment]
cscs-ci run

File: test/unit/test_util_math.cpp [hunk 1]
```diff
@@ -168,7 +169,7 @@ TYPED_TEST(MathUtilTest, ptrdiff_t_Arithmetic_Mul) {
 
   TypeParam a, b;
 
-  if (type_size < arithmetic_size) {
+  if (type_size * 2 < arithmetic_size) {
     a = std::numeric_limits<TypeParam>::max();
     b = std::numeric_limits<TypeParam>::max();
   }
```
Comment (msimberg) [inline test/unit/test_util_math.cpp:175]: I checked that the overflows reported by ubsan do indeed make sense, but please have a look to see if you think the tests are still testing what they should be testing.

**Comment by msimberg** [comment]
cscs-ci run

**Review by albestro — APPROVED** [review]

**Comment by msimberg** [comment]
cscs-ci run

**Pull request merged**